File: generated-code/fsharp/codecraft/TcpReadWrite/Model/DebugInterface/DebugData.fs
#nowarn "0058"

namespace TransGenTest.Model.DebugInterface

open TransGenTest

/// Log some text
type DebugDataLog = {
    /// Text to show
    Text: string;
} with

    /// Write Log to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        let textData : byte[] = System.Text.Encoding.UTF8.GetBytes this.Text
        writer.Write textData.Length
        writer.Write textData

    /// Read Log from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Text = reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString
    }

/// Draw primitives
type DebugDataPrimitives = {
    /// Vertices
    Vertices: Model.DebugInterface.ColoredVertex[];
    /// Primitive type
    PrimitiveType: Model.DebugInterface.PrimitiveType;
} with

    /// Write Primitives to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        writer.Write this.Vertices.Length
        this.Vertices |> Array.iter (fun value ->
            value.writeTo writer )
        writer.Write (int this.PrimitiveType)

    /// Read Primitives from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertices = [|for _ in 1 .. reader.ReadInt32() do
                       yield Model.DebugInterface.ColoredVertex.readFrom reader; |]
        PrimitiveType = reader.ReadInt32() |> enum
    }

/// Draw text
type DebugDataPlacedText = {
    /// Vertex to determine text position and color
    Vertex: Model.DebugInterface.ColoredVertex;
    /// Text
    Text: string;
    /// Text alignment (0 means left, 0.5 means center, 1 means right)
    Alignment: single;
    /// Font size in pixels
    Size: single;
} with

    /// Write PlacedText to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 2
        this.Vertex.writeTo writer
        let textData : byte[] = System.Text.Encoding.UTF8.GetBytes this.Text
        writer.Write textData.Length
        writer.Write textData
        writer.Write this.Alignment
        writer.Write this.Size

    /// Read PlacedText from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Vertex = Model.DebugInterface.ColoredVertex.readFrom reader;
        Text = reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString
        Alignment = reader.ReadSingle()
        Size = reader.ReadSingle()
    }

/// Debug data can be drawn in the app
type DebugData =
    /// Log some text
    | Log of DebugDataLog
    /// Draw primitives
    | Primitives of DebugDataPrimitives
    /// Draw text
    | PlacedText of DebugDataPlacedText
    with

    /// Write DebugData to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | Log value -> value.writeTo writer
            | Primitives value -> value.writeTo writer
            | PlacedText value -> value.writeTo writer

    /// Read DebugData from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> Log (DebugDataLog.readFrom reader)
            | 1 -> Primitives (DebugDataPrimitives.readFrom reader)
            | 2 -> PlacedText (DebugDataPlacedText.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)

File: generated-code/rust/codecraft/FileReadWrite/src/model/debug_interface/debug_data.rs
use super::*;

/// Debug data can be drawn in the app
#[derive(Clone, Debug)]
pub enum DebugData {
    /// Log some text
    Log {
        /// Text to show
        text: String,
    },
    /// Draw primitives
    Primitives {
        /// Vertices
        vertices: Vec<model::debug_interface::ColoredVertex>,
        /// Primitive type
        primitive_type: model::debug_interface::PrimitiveType,
    },
    /// Draw text
    PlacedText {
        /// Vertex to determine text position and color
        vertex: model::debug_interface::ColoredVertex,
        /// Text
        text: String,
        /// Text alignment (0 means left, 0.5 means center, 1 means right)
        alignment: f32,
        /// Font size in pixels
        size: f32,
    },
}

impl trans::Trans for DebugData {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Self::Log {
                text,
            } => {
                <i32 as trans::Trans>::write_to(&0, writer)?;
                text.write_to(writer)?;
            }
            Self::Primitives {
                vertices,
                primitive_type,
            } => {
                <i32 as trans::Trans>::write_to(&1, writer)?;
                vertices.write_to(writer)?;
                primitive_type.write_to(writer)?;
            }
            Self::PlacedText {
                vertex,
                text,
                alignment,
                size,
            } => {
                <i32 as trans::Trans>::write_to(&2, writer)?;
                vertex.write_to(writer)?;
                text.write_to(writer)?;
                alignment.write_to(writer)?;
                size.write_to(writer)?;
            }
        }
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => {
                let text: String = trans::Trans::read_from(reader)?;
                Ok(Self::Log {
                    text,
                })
            }
            1 => {
                let vertices: Vec<model::debug_interface::ColoredVertex> = trans::Trans::read_from(reader)?;
                let primitive_type: model::debug_interface::PrimitiveType = trans::Trans::read_from(reader)?;
                Ok(Self::Primitives {
                    vertices,
                    primitive_type,
                })
            }
            2 => {
                let vertex: model::debug_interface::ColoredVertex = trans::Trans::read_from(reader)?;
                let text: String = trans::Trans::read_from(reader)?;
                let alignment: f32 = trans::Trans::read_from(reader)?;
                let size: f32 = trans::Trans::read_from(reader)?;
                Ok(Self::PlacedText {
                    vertex,
                    text,
                    alignment,
                    size,
                })
            }
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}